File: leetcode/Python/combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        sol, comb = [], []
        def _combine(cur): 
            if len(comb) == k:
                sol.append(comb[:])
                return
            for i in range(cur+1,n+1):
                comb.append(i)
                _combine(i)
                comb.pop()
        _combine(0)
        return sol

File: leetcode/C++/combinations.cpp
class Solution {
public:
    vector<int> comb;
    vector<vector<int>> sol;
    void dfs(int &n, int &k, int cur){
        if(comb.size()==k){
            sol.push_back(comb);
            return;
        }
        for(int i=cur+1; i<=n; ++i) {
            comb.push_back(i);
            dfs(n,k,i);
            comb.pop_back();
        }
    }
    vector<vector<int>> combine(int n, int k) {
        dfs(n,k,0);
        return sol;
    }
};